components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java

```diff
@@ -159,7 +168,38 @@ private ModelSummary buildKMeansModel(long modelID, JavaRDD<Vector> trainingData
 //            if (testingData != null) {
 //                testingData.cache();
 //            }
-            
+
+            // generating data for summary clusters
+            double sampleSize = (double) MLCoreServiceValueHolder.getInstance().getSummaryStatSettings()
+                    .getSampleSize();
+            double sampleFraction = sampleSize / (trainingData.count() - 1);
+            JavaRDD<Vector> sampleData = null;
+            if (sampleFraction >= 1.0) {
+                sampleData = trainingData;
+            }
+            // Use ramdomly selected sample fraction of rows if number of records is > sample fraction
+            else {
+                sampleData = trainingData.sample(false, sampleFraction);
+            }
+
```

Comment:
At this point, please unpersist training data (as it is no more used) and cache sampleData.